libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -26,29 +31,64 @@ import kotlin.coroutines.resume
  * This class is responsible for refreshing the current direction route's traffic.
  * This does not support alternative routes.
  */
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
 internal class RouteRefreshController(
     private val routeRefreshOptions: RouteRefreshOptions,
     private val routeRefresh: RouteRefresh,
     private val currentIndicesProvider: CurrentIndicesProvider,
     private val routeDiffProvider: DirectionsRouteDiffProvider = DirectionsRouteDiffProvider(),
-    private val localDateProvider: () -> Date
+    threadController: ThreadController,
+    private val localDateProvider: () -> Date,
 ) {
 
+    private val mainJobControl = threadController.getMainScopeAndRootJob()
+
+    @RouteRefreshExtra.RouteRefreshState
+    private var state: String = RouteRefreshExtra.REFRESH_STATE_IDLE
+        set(value) {
+            if (field == value) return
+            field = value
+            val _observer = observer ?: return
+            mainJobControl.scope.launch { _observer.onNewState(value) }
+        }
+
+    @Volatile
```

Comment:
Can it be accessed from different threads? It looks like it's only main now.

Reply:
we don't obligate users to set observer via main thread

Reply:
`MapboxNavigation` class is marked with `@UiThread` annotation.

Reply:
you're right 👍 gonna remove volatile annotation